client/src/main/java/crawlercommons/urlfrontier/client/ListURLs.java

```diff
@@ -0,0 +1,137 @@
+            required = false,
+            paramLabel = "STRING",
+            description = "key to use to target a specific queue")
+    private String key;
+
+    @Option(
+            names = {"-o", "--output"},
+            defaultValue = "",
+            paramLabel = "STRING",
+            description = "output file to dump all the URLs")
+    private String output;
+
+    @Option(
+            names = {"-c", "--crawlID"},
+            defaultValue = "DEFAULT",
+            paramLabel = "STRING",
+            description = "crawl to get the queues for")
+    private String crawl;
+
+    @Option(
+            names = {"-l", "--local"},
+            defaultValue = "false",
+            paramLabel = "BOOLEAN",
+            description =
+                    "restricts the scope to this frontier instance instead of aggregating over the cluster")
+    private Boolean local;
+
+    @Option(
+            names = {"-p", "--parsedate"},
+            defaultValue = "false",
+            description = {
+                "Print the refetch date in local time zone",
+                "By default, time is in UTC seconds since the Unix epoch"
+            })
+    private boolean parse;
+
+    // Use the system default time zone
+    private ZoneId zoneId = ZoneId.systemDefault();
+
+    @Override
+    public void run() {
+
+        Builder builder = crawlercommons.urlfrontier.Urlfrontier.ListUrlParams.newBuilder();
+        builder.setLocal(local);
+        builder.setKey(key);
+        builder.setSize(maxNumURLs);
+        builder.setStart(start);
+        builder.setCrawlID(crawl);
+
+        PrintStream outstream = null;
+        if (output.length() > 0) {
+            File f = new File(output);
+            f.delete();
+            try {
+                outstream = new PrintStream(f, Charset.defaultCharset());
+            } catch (IOException e) {
+                e.printStackTrace(System.err);
+                return;
+            }
+        } else {
+            outstream = System.out;
+        }
+
+        ManagedChannel channel =
+                ManagedChannelBuilder.forAddress(parent.hostname, parent.port)
+                        .usePlaintext()
+                        .build();
+        URLFrontierBlockingStub blockingFrontier = URLFrontierGrpc.newBlockingStub(channel);
+
+        Iterator<URLItem> it = blockingFrontier.listURLs(builder.build());
+        while (it.hasNext()) {
+
+            URLItem item = it.next();
+
+            String fetchDate;
+            if (parse) {
+                Instant instant = Instant.ofEpochSecond(item.getKnown().getRefetchableFromDate());
+                LocalDateTime localDate = instant.atZone(zoneId).toLocalDateTime();
+                fetchDate = localDate.toString();
+            } else {
+                fetchDate = String.valueOf(item.getKnown().getRefetchableFromDate());
+            }
+
+            outstream.println(item.getKnown().getInfo().getUrl() + ";" + fetchDate);
```

Comment:
it would be good to be able to see any metadata associated with the URL. 
IIRC it is possible to de-ser in JSON pretty easily, see [PutURL clients](https://github.com/crawler-commons/url-frontier/blob/08f09c37a64290114ba102a108111c2fcfffe8e3/client/src/main/java/crawlercommons/urlfrontier/client/PutURLs.java#L166). It would be good to have a way of specifying the output format between JSON or char separated. 
Maybe we could have a utility class to share the logic of reading to / from various formats? This could be done later

Reply:
> it would be good to be able to see any metadata associated with the URL.

I see one problem though, whether it is in MemoryFrontier or in RocksDB, we lose the Metadata once a URL is completed

Reply:
> > it would be good to be able to see any metadata associated with the URL.
> 
> I see one problem though, whether it is in MemoryFrontier or in RocksDB, we lose the Metadata once a URL is completed

Ah, maybe something to fix separately. I suppose we could still display the metadata where possible